### PR TITLE
add reconciliation to the engine

### PR DESCRIPTION
**Pull request by jhsinger-klotho**
follows https://github.com/klothoplatform/klotho/pull/683

Creates a dag for data flow, used for another visualization 

Adds reconciliation to the engine so we can delete resources

### Standard checks

- **Unit tests**: Any special considerations?
- **Docs**: Do we need to update any docs, internal or public?
- **Backwards compatibility**: Will this break existing apps? If so, what would be the extra work required to keep them working?

### CODE REVIEW

File: pkg/engine/dataflow.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package engine
+
+import (
+	"github.com/klothoplatform/klotho/pkg/collectionutil"
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/provider/aws/resources"
+	"go.uber.org/zap"
+)
+
+func (e *Engine) GetDataFlowDag() *core.ResourceGraph {
```
Comment (DavidSeptimus-Klotho) [inline pkg/engine/dataflow.go:10]: This is neat.

File: pkg/knowledge_base/edge_kb.go [hunk 2]
```diff
@@ -402,3 +403,41 @@ func (kb EdgeKB) findPathsInGraph(source, dest core.Resource, stack []graph.Edge
 	delete(visited, source)
 	return result
 }
+
+func (kb EdgeKB) GetTrueUpstream(source core.Resource, dag *core.ResourceGraph) []core.Resource {
```
Comment (DavidSeptimus-Klotho) [inline pkg/knowledge_base/edge_kb.go:407]: Can you add docs for `GetTrueUpstream` and `GetTrueDownstream` to make their purpose clear?

File: pkg/visualizer/file.go [hunk 3]
```diff
@@ -49,7 +50,7 @@ type (
 )
 
 func (f *File) Path() string {
-	return "topology.yaml"
+	return fmt.Sprintf("%stopology.yaml", f.PathPrefix)
```
Comment (DavidSeptimus-Klotho) [inline pkg/visualizer/file.go:53]: nit: should we be `filepath.Join()` here instead of a string format?

File: pkg/visualizer/plugin.go [hunk 4]
```diff
@@ -106,6 +106,46 @@ func (p Plugin) Translate(dag *core.ResourceGraph) ([]core.File, error) {
...
 }
 
+// Translate implements compiler.IaCPlugin - although it's not strictly an IaC plugin, it uses the same API
+func (p Plugin) Generate(dag *core.ResourceGraph, filenamePrefix string) ([]core.File, error) {
+	api := visApi{client: p.Client}
+
+	if validateTypes {
+		types := TypesChecker{DAG: dag}
+
+		if resp, err := api.request(http.MethodGet, `validate-types`, `application/text`, ``, types); err != nil {
```
Comment (DavidSeptimus-Klotho) [inline pkg/visualizer/plugin.go:116]: We might want to disable this warning since the new visualizer doesn't have this capability just yet.

**Review by DavidSeptimus-Klotho — APPROVED** [review]

File: pkg/knowledge_base/edge_kb.go (hunk 2 above)
Comment (jhsinger-klotho) [reply pkg/knowledge_base/edge_kb.go:407]: yeah going to add docs to the new pr

File: pkg/visualizer/file.go (hunk 3 above)
Comment (jhsinger-klotho) [reply pkg/visualizer/file.go:53]: the path prefix is really just a filename prefix so i will rename

**Pull request merged**